### PR TITLE
Clear reference to training loss at the end of train step

### LINKED ISSUE #9343: OOM after a few iterations on my model
Hello,

I have been trying to debug OOM after a few iterations on my model.
After investigation I saw that in loops/batch/training_batch_loop.py there is this piece of code (L235):
`result = self.training_step_and_backward(split_batch, batch_idx, opt_idx, optimizer, hiddens)`
`  if result is not None:`
`     return_result.update(result)`
`     return return_result.loss`
     
What I see is that return_result will keep the computation graph as it contains the loss. So I wonder what is the role of this variable? Also, where is the graph released, I could no go any further than "_training_step_and_backward_closure"? I don't understand why my model runs fine for a few iterations then there is some increase in memory.

_Originally posted by @Jovp in https://github.com/PyTorchLightning/pytorch-lightning/discussions/9332_

### PR DESCRIPTION
**Pull request by martenlienen**
## What does this PR do?

Without clearing this reference, the loss tensor stays live through the next training
step. This can be a problem for memory intensive models that produce very deep backward
graphs such as neural ODEs. For these models, keeping the backward graph of the previous
loss in memory can lead to OOM errors in the next training step even though the step might
have succeeded if we had cleared (and thus GC'd) the previous backward graph.

Fixes #9343

### Does your PR introduce any breaking changes? If yes, please list them.

No

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 1]
```diff
@@ -137,6 +137,10 @@ def advance(self, batch, batch_idx):
             if result:
                 self.batch_outputs[0].append(deepcopy(result.result_collection))
 
+            # clear reference to this step's training loss so that it can be garbage
+            # collected before the next training step
+            trainer._results.minimize = None
```
Comment (rohitgr7) [inline pytorch_lightning/loops/batch/training_batch_loop.py:142]: shouldn't it go inside the `automatic_optimization` block right above?

Comment (martenlienen) [reply pytorch_lightning/loops/batch/training_batch_loop.py]: Yes, of course

File: pytorch_lightning/loops/batch/training_batch_loop.py [hunk 2]
```diff
@@ -137,6 +137,10 @@ def advance(self, batch, batch_idx):
             if result:
                 self.batch_outputs[0].append(deepcopy(result.result_collection))
 
+                # clear reference to this step's training loss so that it can be garbage
+                # collected before the next training step
+                trainer._results.minimize = None
```
Comment (carmocca) [inline pytorch_lightning/loops/batch/training_batch_loop.py:142]: Is this only a problem with manual optimization?

This should probably be moved to `LoggerConnector.on_train_split_start`.

cc @tchaton this is related to avoiding the `ResultCollection` as a vehicle to pass data between loops and iterations as the state is harder to manage (as we saw with deepcopy)

Comment (martenlienen) [reply pytorch_lightning/loops/batch/training_batch_loop.py:142]: No, it happens with automatic optimization. My use case is a bit particular. I am training neural ODEs that can have varying memory requirements per batch and sometimes the memory usage spikes causing a CUDA OOM and aborting the training. To alleviate that a bit and filter our problematic batches, my module looks basically like the following
```python
import pytorch_lightning as pl
from pytorch_lightning.utilities.memory import (
    garbage_collection_cuda,
    is_cuda_out_of_memory,
)

class ExampleModule(pl.LightningModule):
    def training_step(self, *args):
        try:
            return { "loss": ... }
        except RuntimeError as e:
            if is_cuda_out_of_memory(e):
                return None
            else:
                raise

    def backward(self, *args, **kwargs):
        try:
            return super().backward(*args, **kwargs)
        except RuntimeError as e:
            if is_cuda_out_of_memory(e):
                return None
            else:
                raise
```
What can happen then is that a CUDA OOM occurs during the backward pass. In that case the context tensors of the backwards graph do not get freed. Now, if I have a loss with a huge backwards graph (let's say 99% of CUDA memory), `backward` will surely fail, leaving the whole graph in memory. That graph is still attached to the loss that is stored in `trainer._results.minimize`. Because that reference keeps the graph and memory alive throughout the next training step leaving only 1% of CUDA memory, no forward pass will ever succeed again and `trainer._results.minimize` will never be overwritten. This PR fixes that by clearing that reference as soon as possible.

All in all, I agree that the root cause is this mutable `trainer._results` object that is modified and copied for some reason instead of just creating new instances.

Comment (tchaton) [reply pytorch_lightning/loops/batch/training_batch_loop.py]: Yes, we need to refactor the code, but we can merge this one as a hotfix I guess.

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/loops/batch/training_batch_loop.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/loops/batch/training_batch_loop.py]: Still as, I said:

> This should probably be moved to LoggerConnector.on_train_split_start.

So it gets freed before

**Review by carmocca — APPROVED** [review]

File: tests/trainer/loops/test_training_loop.py [hunk 3]
```diff
@@ -190,3 +190,14 @@ def training_epoch_end(self, outputs) -> None:
     trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=2)
     trainer.fit(model)
     assert model.on_train_batch_end_called == 2
+
+
+def test_batch_loop_releases_loss(tmpdir):
+    class TestModel(BoringModel):
```
Comment (awaelchli) [inline tests/trainer/loops/test_training_loop.py:196]:
```suggestion
def test_batch_loop_releases_loss(tmpdir):
    """Test that loss/graph is released and that it can be garbage collected before the next training step"""
    class TestModel(BoringModel):
```

**Review by awaelchli — APPROVED** [review]
